Wangscape/image/Image.cpp

```diff
@@ -30,3 +63,254 @@ ImageGrey imageToGreyscale(const ImageColour & image)
+    const auto nz_size = (nz_max - nz_min + 1).eval();
+    return sf::IntRect(nz_min(1), nz_min(0), nz_size(1), nz_size(0));
+}
+
+sf::IntRect boxUnion(sf::IntRect a, sf::IntRect b)
+{
+    arma::imat origins = {{a.left, a.top}, {b.left, b.top}};
+    arma::imat sizes = {{a.width, a.height}, {b.width, b.height}};
+    auto ends = (origins + sizes - 1).eval();
+    auto origin_min = arma::min(origins, 0).eval();
+    auto end_max = arma::max(ends, 0).eval();
+    auto size_max = (end_max - origin_min + 1).eval();
+    return sf::IntRect(origin_min(0), origin_min(1),
+                       size_max(0), size_max(1));
+}
+
+ImageGrey32 distances(const ImageGrey & traversable, const ImageGrey & targets, bool use_diagonals)
+{
+    if (!isBinary(traversable))
+        throw std::runtime_error("distances() requires binary traversable mask");
+    if (!isBinary(targets))
+        throw std::runtime_error("distances() requires binary targets mask");
+    if(!isNonzero(traversable))
+        throw std::runtime_error("distances() requires nonzero traversable mask");
+    if (!isNonzero(targets))
+        throw std::runtime_error("distances() requires nonzero targets mask");
+    if(!isConnected(traversable))
+        throw std::runtime_error("distances() requires connected traversable mask");
+    const auto shape = makeUVec(arma::size(traversable));
+    ImageGrey32 distance_field(arma::size(traversable));
+    const arma::u32 infinity = std::numeric_limits<arma::u32>::max();
+    distance_field.fill(infinity);
+    const auto target_coordinates = arma::ind2sub(arma::size(targets), arma::find(targets)).eval();
+    std::queue<IVec> border;
+    const auto& steps = use_diagonals ? ORTHODIAGONAL_STEPS : ORTHOGONAL_STEPS;
+    const auto in_bounds = [&shape](IVec point)
+    {
+        if (point.x() < 0 || point.y() < 0)
+            return false;
+        if ((unsigned int)point.x() >= shape.x() || (unsigned int)point.y() >= shape.y())
+            return false;
+        return true;
+    };
+    for (unsigned int i = 0; i < target_coordinates.n_cols; i++)
+    {
+        IVec target{int(target_coordinates(1, i)), int(target_coordinates(0, i))};
+        distance_field(target.y(), target.x()) = 0;
+        border.push(target);
+    }
+    while (!border.empty())
+    {
+        IVec point = border.front();
+        border.pop();
+        arma::u32 distance = distance_field(point.y(), point.x()) + 1;
+        for (const auto step : steps)
+        {
+            IVec new_point = point + step;
+            if (!in_bounds(new_point))
+                continue;
+            if (!traversable(new_point.y(), new_point.x()))
+                continue;
+            if (distance_field(new_point.y(), new_point.x()) == infinity)
+            {
+                border.push(new_point);
+                distance_field(new_point.y(), new_point.x()) = distance;
+            }
+        }
+    }
+    return distance_field;
+}
```

Comment:
Add newlines to separate parts of this function. It will be much easier to read :).